lib/tools/adb-commands.js

```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
+  await this.keyevent(26);
+};
+
+methods.back = async function () {
+  log.debug("Pressing the BACK button");
+  await this.keyevent(4);
+};
+
+methods.goToHome = async function () {
+  log.debug("Pressing the HOME button");
+  await this.keyevent(3);
+};
+
+methods.isScreenLocked = async function () {
+  let stdout = await this.shell(['dumpsys', 'window']);
+  if (process.env.APPIUM_LOG_DUMPSYS) {
+    // optional debugging
+    // if the method is not working, turn it on and send us the output
+    let dumpsysFile = path.resolve(process.cwd(), "dumpsys.log");
+    log.debug(`Writing dumpsys output to ${dumpsysFile}`);
+    await fs.writeFile(dumpsysFile, stdout);
+  }
+  return (isShowingLockscreen(stdout) || isCurrentFocusOnKeyguard(stdout) ||
+          !isScreenOnFully(stdout));
+};
+
+methods.isSoftKeyboardPresent = async function () {
+  try {
+    let stdout = await this.shell(['dumpsys', 'input_method']);
+    let isKeyboardShown = false,
+        canCloseKeyboard = false,
+        inputShownMatch = /mInputShown=\w+/gi.exec(stdout);
+    if (inputShownMatch && inputShownMatch[0]) {
+      isKeyboardShown = inputShownMatch[0].split('=')[1] === 'true';
+      let isInputViewShownMatch = /mIsInputViewShown=\w+/gi.exec(stdout);
+      if (isInputViewShownMatch && isInputViewShownMatch[0]) {
+        canCloseKeyboard = isInputViewShownMatch[0].split('=')[1] === 'true';
+      }
+    }
+    return {isKeyboardShown:isKeyboardShown, canCloseKeyboard:canCloseKeyboard};
+  } catch (e) {
+    log.errorAndThrow(`Error finding softkeyboard. Original error: ${e.message}`);
+  }
+};
+
+methods.sendTelnetCommand = async function (command) {
+  log.debug(`Sending telnet command to device: ${command}`);
+  let port = await this.getEmulatorPort();
+  return new Promise(function(resolve, reject) {
+    let conn = net.createConnection(port, 'localhost'),
+        connected = false,
+        readyRegex = /^OK$/m,
+        dataStream = "",
+        res = null;
+    conn.on('connect', function () {
+      log.debug("Socket connection to device created");
+    });
+    conn.on('data', function (data) {
+      data = data.toString('utf8');
+      if (!connected) {
+        if (readyRegex.test(data)) {
+          connected = true;
+          log.debug("Socket connection to device ready");
+          conn.write(command + "\n");
+        }
+      } else {
+        dataStream += data;
+        if (readyRegex.test(data)) {
+          res = dataStream.replace(readyRegex, "").trim();
+          log.debug(`Telnet command got response: ${res}`);
+          conn.write("quit\n");
+        }
+      }
+    });
+    conn.on('close', function () {
+      if (res === null) {
+        reject(new Error("Never got a response from command"));
+      } else {
+        resolve(res);
+      }
+    });
+  });
+};
+
+methods.isAirplaneModeOn = async function () {
+  let stdout = this.shell(['settings', 'get', 'global', 'airplane_mode_on']);
```

Comment:
don't we need to `await this.shell`? and if so, shouldn't there be a test that catches this?